### PR TITLE
Openid connect new o auth policy fields

### LINKED ISSUE #1277: Openid connect new OAuthPolicy fields
This change is going to introduce new fields and CRUD operations into OAuthPolicy and Scope for openid connect.

### PR DESCRIPTION
**Pull request by mrafiei**
Resolves #1277

### CODE REVIEW

**Comment by mrioan** [comment]
hello @mrafiei can you please update #1277 in order to have a better understanding of what this issue is for? Otherwise it is difficult for us to know how to review this PR. Thanks!

File: api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicy.java [hunk 1]
```diff
@@ -72,6 +84,71 @@
...
+     *
+     * @return this instance for method chaining.
+     */
+    OAuthPolicy setIdTokenTtl(String idTokenTtl);
+
+    /**
+     * Creates a new {@link Scope} assigned to this oauthPolicy in the Stormpath server and returns the created resource.
+     * The scope is used for openid connect flows.
+     *
+     * @param scope {@link Scope} pojo to hold necessary data to send to the back- end to create a {@link Scope}.
```
Comment (mrioan) [inline api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicy.java]: back-end

File: api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicy.java [hunk 2]
```diff
@@ -72,6 +84,71 @@
...
+     * Returns a paginated list of all the scopes that belong to the oAuthPolicy.
+     *
+     * @return a paginated list of all the oAuthPolicy's scopes.
+     *
+     * @since 1.6.0
+     */
+    ScopeList getScopes();
+
+    /**
+     * Returns access token attribute mappings.
```
Comment (mrioan) [inline api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicy.java]: What is an "attribute mapping"? Can you please describe that a little bit? thanks

File: api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicy.java [hunk 3]
```diff
@@ -72,6 +84,71 @@
...
+    OAuthPolicy setAccessTokenAttributeMap(Map<String,String> accessTokenAttributeMap);
+
+    /**
+     * Returns id token attribute mappings.
+     *
+     * @return id token attribute mappings.
+     *
+     * @since 1.6.0
+     */
+    Map<String,String> getIdTokenAttributeMap();
```
Comment (mrioan) [inline api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicy.java]: Can you please add the description of what an attribute mapping is here? thanks

File: api/src/main/java/com/stormpath/sdk/oauth/openidconnect/Scope.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+import com.stormpath.sdk.resource.Resource;
+import com.stormpath.sdk.resource.Saveable;
+
+import java.util.Map;
+
+/**
+ * A Scope resource is used to configure different openid connect scopes
+ * with an {@link OAuthPolicy OAuthPolicy}
+ *
+ * @since 1.0.RC7
```
Comment (mrioan) [inline api/src/main/java/com/stormpath/sdk/oauth/openidconnect/Scope.java]: 1.6.0

File: extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/application/OAuthPolicyIT.groovy [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+    @Test
+    void testPolicyExpandOnScopes(){
+        def app = createTempApp()
+        def policyHref = app.getOAuthPolicy().href+"?expand=scopes"
+
+        def oauthPolicy = client.getResource(policyHref, OAuthPolicy.class)
+        assertNotNull oauthPolicy
+        assertNotNull oauthPolicy.href
+
+        assertEquals(oauthPolicy.scopes.size, 3)
```
Comment (mrioan) [inline extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/application/OAuthPolicyIT.groovy]: this test does not properly validate that the scopes have been expanded.

Regardless of whether the scope is expanded or not the `oauthPolicy.scopes.size` will return the number of scopes, but you will not know if there was a call to the backend or not. 

The proper way to validate the the scopes are expanded is to manually read the internal values of the application without using its public "getScopes" method. For example, check `TenantIT#testTenantExpansionWithCache`

You will see something like this `Map tenantProperties = getValue(AbstractResource, retrieved, "properties")`
 -> this map will contain the local properties. Now, you can continue asserting its current state  knowing that the backend will not be hit. If the data that you are looking for is not there then the resource has not been properly expanded (therefore this test will be working properly)

File: impl/src/main/java/com/stormpath/sdk/impl/ds/cache/WriteCacheFilter.java [hunk 6]
```diff
@@ -84,14 +77,6 @@ public WriteCacheFilter(BaseUrlResolver baseUrlResolver, CacheResolver cacheReso
         this.referenceFactory = referenceFactory;
         this.cacheMapInitializer = new DefaultCacheMapInitializer();
         this.baseUrlResolver = baseUrlResolver;
-        this.webConfigurationMaps = new HashSet<>();
-        this.webConfigurationMaps.add("oauth2");
-        this.webConfigurationMaps.add("register");
-        this.webConfigurationMaps.add("login");
```
Comment (mrioan) [inline impl/src/main/java/com/stormpath/sdk/impl/ds/cache/WriteCacheFilter.java]: This change comes form a commit that does not belong to this PR. I think this commit is already in master (at least it is released in `1.5.2`, so should be in `master` already), can you please make sure that it is not showing up here? (action: pull master)

File: api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicy.java [hunk 7]
```diff
@@ -43,6 +48,13 @@
     String getRefreshTokenTtl();
 
     /**
+     * Returns the Time To Live for the id tokens created for the parent {@link Application Application} expressed in a period of time format, for example: PT1H.
+     *
+     * @return the String representation of the Time To Live for the id tokens created for the parent {@link Application Application}
+     */
+    String getIdTokenTtl();
```
Comment (mrafiei) [inline api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicy.java]: @mrioan Thanks for the review.
Id token, Access Token and Refresh Token are well known idioms in OAuth2 and openid connect world. Whoever is using this code is trying to set or consume openid connect configuration and would certainly know what an id token is.
Therefore I tried keeping the javodoc for it similar to the existing access token and refresh token.

File: api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicy.java [hunk 8]
```diff
@@ -72,6 +84,71 @@
     OAuthPolicy setRefreshTokenTtl(String refreshTokenTtl);
 
     /**
+     * Sets the Time To Live for the id tokens created for the parent {@link Application Application} expressed in a period of time format, for example: PT1H.
+     *
+     * @return this instance for method chaining.
+     */
+    OAuthPolicy setIdTokenTtl(String idTokenTtl);
```
Comment (mrafiei) [inline api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicy.java]: Added!

File: api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicy.java (hunk 1 above)
Comment (mrafiei) [inline api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicy.java]: Corrected!

File: api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicy.java (hunk 2 above)
Comment (mrafiei) [inline api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicy.java]: Elaborated more on that!

File: api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicy.java (hunk 3 above)
Comment (mrafiei) [inline api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicy.java]: Elaborated more on that!

File: extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/application/OAuthPolicyIT.groovy (hunk 5 above)
Comment (mrafiei) [inline extensions/httpclient/src/test/groovy/com/stormpath/sdk/impl/application/OAuthPolicyIT.groovy]: Added "OAuthPolicyOptions", "OAuthPolicies" and made it similar to "TenantIT#testTenantExpansionWithCache"

File: api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicy.java (hunk 7 above)
Comment (mrioan) [inline api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicy.java]: OK, thanks!

File: api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicy.java [hunk 9]
```diff
@@ -72,6 +85,80 @@
...
+     * Authorization server would then add all these mappings as part of the access token upon its generation.
+     * @return this instance for method chaining.
+     *
+     * @since 1.6.0
+     */
+    OAuthPolicy setAccessTokenAttributeMap(Map<String,String> accessTokenAttributeMap);
+
+    /**
+     * Returns id token attribute mappings.
+     * Open Id provider (OP) would enter any custom mappings used for their internal purposes in this map.
```
Comment (mrioan) [inline api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicy.java]: please surround each new paragraph with `<p> .... </p>`

File: api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicyOptions.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+package com.stormpath.sdk.oauth;
+
+import com.stormpath.sdk.query.Options;
+
+/**
+ * @since 1.6.0
+ */
+public interface OAuthPolicyOptions<T> extends Options {
+
+    T withScopes();
```
Comment (mrioan) [inline api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicyOptions.java]: Please add javadocs to these methods

File: api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicy.java (hunk 9 above)
Comment (mrafiei) [inline api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicy.java]: Added!

File: api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicyOptions.java (hunk 10 above)
Comment (mrafiei) [inline api/src/main/java/com/stormpath/sdk/oauth/OAuthPolicyOptions.java]: Added!

**Comment by mrafiei** [comment]
Thanks for the review. I have done the pending comments.
This would only pass Travis once the feature is fully available in IAM. Will merge it once that's the case.
